sw/host/spiflash/ftdi_spi_interface.cc

```diff
@@ -88,7 +78,24 @@ FtdiSpiInterface::~FtdiSpiInterface() {
 }
 
 bool FtdiSpiInterface::Init() {
-  struct mpsse_context *ctx = MPSSE(SPI0, kFrequency, MSB);
+  if (!options_.device_serial_number.empty() &&
+      (options_.device_vendor_id == 0 || options_.device_product_id == 0)) {
+    std::cerr << "FTDI device serial number requires the vendor_id and product "
+                 "id to be set."
+              << std::endl;
+    return false;
+  }
+
+  struct mpsse_context *ctx = nullptr;
+  if (options_.device_vendor_id == 0 || options_.device_product_id == 0) {
+    ctx = MPSSE(SPI0, options_.spi_frequency, MSB);
+  } else {
+    const char *serial_number = options_.device_serial_number.empty()
+                                    ? nullptr
+                                    : options_.device_serial_number.c_str();
+    ctx = Open(options_.device_vendor_id, options_.device_product_id, SPI0,
+               options_.spi_frequency, MSB, IFACE_A, nullptr, serial_number);
```

Comment:
Would you mind adding /* arg = */ annotations to function arguments here that aren't obvious? E.g. SPI0, MSB, etc.

Reply:
Done.